File: web/src/lib/api/governance.ts
import { apiFetch } from "./client"

export interface Approval {
  id: number
  gate_type: string
  status: string
  pipeline_id: string | null
  payload_ref: Record<string, unknown> | null
  impact_summary: Record<string, unknown> | null
  submitted_at: string | null
  decided_at: string | null
  decided_by: number | null
  decision_reason: string | null
  expires_at: string | null
}

export interface ApprovalListResponse {
  approvals: Approval[]
}

export async function getApprovals(status?: string): Promise<ApprovalListResponse> {
  const params = status ? `?status=${status}` : ""
  return apiFetch<ApprovalListResponse>(`/api/v1/admin/approvals${params}`)
}

export async function approveApproval(id: number, reason?: string): Promise<{ status: string }> {
  return apiFetch(`/api/v1/admin/approvals/${id}/approve`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  })
}

export async function rejectApproval(id: number, reason?: string): Promise<{ status: string }> {
  return apiFetch(`/api/v1/admin/approvals/${id}/reject`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  })
}

export interface DashboardResponse {
  pending_count: number
  avg_approval_latency_hours: number | null
  rejection_rate: number | null
  recent_anomalies: unknown[]
}

export interface TransparencyResponse {
  oversight_levels: Record<string, string[]>
  last_approvals: Record<string, { decided_at: string; status: string }>
  pipeline_health: { status: string; last_successful_run: string | null }
}

export async function getDashboard(): Promise<DashboardResponse> {
  return apiFetch<DashboardResponse>("/api/v1/admin/governance/dashboard")
}

export async function getTransparency(): Promise<TransparencyResponse> {
  return apiFetch<TransparencyResponse>("/api/v1/governance/transparency")
}

export interface GovernanceEvent {
  id: number
  event_type: string
  source: string
  detail: Record<string, unknown> | null
  created_at: string | null
}

export interface EventListResponse {
  events: GovernanceEvent[]
  total: number
}

export async function getEvents(
  options?: { event_type?: string; limit?: number; offset?: number }
): Promise<EventListResponse> {
  const params = new URLSearchParams()
  if (options?.event_type) params.set("event_type", options.event_type)
  if (options?.limit) params.set("limit", String(options.limit))
  if (options?.offset) params.set("offset", String(options.offset))
  const query = params.toString()
  return apiFetch<EventListResponse>(`/api/v1/admin/governance/events${query ? `?${query}` : ""}`)
}

File: api/src/margin_api/schemas/governance.py
"""Pydantic schemas for governance admin endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class ApprovalSummary(BaseModel):
    id: int
    gate_type: str
    status: str
    pipeline_id: str | None = None
    payload_ref: dict | None = None
    impact_summary: dict | None = None
    submitted_at: datetime | None = None
    decided_at: datetime | None = None
    decided_by: int | None = None
    decision_reason: str | None = None
    expires_at: datetime | None = None


class ApprovalListResponse(BaseModel):
    approvals: list[ApprovalSummary]


class ApprovalDecisionRequest(BaseModel):
    reason: str | None = None


class GovernanceEventResponse(BaseModel):
    id: int
    event_type: str
    source: str
    detail: dict | None = None
    created_at: datetime | None = None


class GovernanceEventListResponse(BaseModel):
    events: list[GovernanceEventResponse]
    total: int


class GovernanceDashboardResponse(BaseModel):
    pending_count: int
    avg_approval_latency_hours: float | None = None
    rejection_rate: float | None = None
    recent_anomalies: list[dict] = []


class TransparencyResponse(BaseModel):
    oversight_levels: dict
    last_approvals: dict
    pipeline_health: dict


class GovernanceConfigResponse(BaseModel):
    config_key: str
    config_value: dict
    description: str
    is_default: bool
    updated_at: datetime | None = None


class GovernanceConfigUpdate(BaseModel):
    config_value: dict


class GovernanceConfigListResponse(BaseModel):
    configs: list[GovernanceConfigResponse]
